# Question2: Debt Prediction
## Fabio Carrasco

#### Read the dataset file “Credit.csv” (from github using the following command), and assign it to a Pandas DataFrame:

In [71]:
import pandas as pd

In [72]:
credits_df  = pd.read_csv("https://github.com/mpourhoma/CS4661/raw/master/Credit.csv")

#### Check out the dataset. The “Credit” dataset includes the “balance” column (average credit card debt for a number of individuals) as target, as well as several features: age, cards (number of credit cards), education (years of education), income (in thousands of dollars), limit (credit limit), marital status, and rating (credit rating).

In [73]:
credits_df[0::10]

Income  Limit  Rating  Cards  Age  Education  Married  Balance
0     14.891   3606     283      2   34         11        1      333
10    63.095   8117     589      4   30         14        1     1407
20    17.700   2860     235      4   63         16        0       89
30    34.142   5666     413      4   47          5        1      863
40    34.950   3327     253      3   54         14        0       50
50    36.362   5183     376      3   49         15        1      654
60    35.510   5198     364      2   35         20        0      631
70    24.889   3954     318      4   75         12        1      357
80    27.369   3449     288      3   40          9        1      162
90    20.191   5767     431      4   42         16        1     1023
100   21.153   3736     256      1   41         11        0      298
110   34.537   3271     250      3   57         17        1       47
120   27.241   1402     128      2   67         15        1        0
130   23.793   3821     281      4   56         12        1      868
140   39.831   6045     459      3   32         12        1     1425
150   63.931   5728     435      3   28         14        1      581
160   12.238   4865     381      5   67         11        0      836
170   34.772   2021     167      3   57          9        0        0
180   10.635   3584     294      5   69         16        1      423
190   55.187   5352     385      4   50         17        1      538
200   23.949   5343     383      2   40         18        1      829
210   24.543   3206     243      2   62         12        1       95
220   44.847   5765     437      3   53         13        0     1246
230   33.214   5137     387      3   59          9        0      661
240   29.705   3351     262      5   71         14        1      148
250   10.363   2430     191      2   47         18        1        0
260   67.937   5184     383      4   63         12        1      345
270   15.866   3085     217      1   39         13        0      136
280   53.401   5319     377      3   35         12        0      541
290   26.370   3235     268      5   78         11        1      159
300   21.786   4632     355      1   50         17        1      580
310   33.694   4891     369      1   52         16        0     1036
320   16.279   1160     126      3   78         13        1        5
330   20.405   4543     329      2   72         17        0     1054
340   27.794   3807     301      4   35          8        1      320
350   30.002   1561     155      4   70         13        1        0
360   53.566   5891     434      4   82         10        0      712
370   35.610   6135     466      4   40         12        0      992
380  115.123   7760     538      3   83         14        0      661
390  135.118  10578     747      3   81         15        1     1393

#### Generate the feature matrix and target vector (target is “balance” in this dataset). Then, normalize (scale) the features (note: don’t normalize the target vector!). To normalize the data, you can simply use scale(X)  from sklearn.

In [74]:
X = credits_df.loc[:, credits_df.columns != 'Balance']
y = credits_df['Balance']

X.head()

Income  Limit  Rating  Cards  Age  Education  Married
0   14.891   3606     283      2   34         11        1
1  106.025   6645     483      3   82         15        1
2  104.593   7075     514      4   71         11        0
3  148.924   9504     681      3   36         11        0
4   55.882   4897     357      2   68         16        1

#### Split the dataset into testing and training sets with the following parameters: test_size=0.24, random_state=9.

In [75]:
from sklearn.model_selection import train_test_split

In [76]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.24, 
                                                    random_state=9)

#### Use Linear Regression to train a linear model on the training set. Check the coefficients of the linear regression model. Which feature is the most important? Which feature is the least important?

In [77]:
# The following line will import LinearRegression "Class"

from sklearn.linear_model import LinearRegression

In [78]:
# In the following line, "my_linreg" is instantiated as an "object" of LinearRegression "class". 

my_linreg = LinearRegression()

# fitting the model to the training data:
my_linreg.fit(X_train, y_train)

# printing Theta0 using attribute "intercept_":
print(my_linreg.intercept_)

# printing [Theta1, Theta2, Theta3] using attribute "coef_":
print(my_linreg.coef_)

-497.0869076583074
[ -7.5197188    0.15640406   1.6355104   18.19349659  -1.00062374
   3.73501837 -20.17447952]


#### The least important feature is Marriage. The most important is Cards.

#### Predict “balance” for the users in testing set. Then, compare the predicted balance with the actual balance by calculating and reporting the RMSE (as we saw in lab tutorial 4).

In [79]:
# make predictions on the testing set
y_prediction = my_linreg.predict(X_test)

print(y_prediction)

[ 316.89813114  677.5093907   489.03669272  -49.82989971  706.42913216
  864.96954115  897.65752581  108.20297732 -213.73474279  834.94198677
   11.05109139 -227.34447869  472.01450822    4.54554458 -228.17349827
  956.21537845  896.91570524  821.97824267   18.37144232  857.31517013
 1028.30326279  699.22168497 1176.9426756   666.28578465  643.94064891
  690.63628696  589.35273903  -29.37237394  575.15444917  406.78468032
  845.81453327  828.30862904   82.2376206   953.40592514 -184.74828227
  525.84335889 1045.19205022  535.04104472  107.26753118 -172.8968111
  497.84037664 1155.35533939  429.08674601  403.49872552  143.72087313
  764.78737012  431.24114308 1304.79656588  475.00132057   95.75404622
 -215.69000296    5.62900657  313.88797405  859.9328932   792.27147016
 1029.90210771 1500.40180297 1078.48617624  713.37981473 1072.97948262
  -64.12187718  488.60063387   92.3601712   440.414465    277.13284667
 1075.41508762  258.10181743  773.89657462  278.12447916  409.51029492
  335.5

In [80]:
from sklearn import metrics
import numpy as np

# Calculating "Mean Square Error" (MSE):
mse = metrics.mean_squared_error(y_test, y_prediction)

# Using numpy sqrt function to take the square root and calculate "Root Mean Square Error" (RMSE)
rmse = np.sqrt(mse)

print(rmse)

143.24649739643215


#### The RSME is 143.25

#### Now, use 10-fold Cross-Validation to evaluate the performance of a linear regression in predicting the balance. Thus, rather than splitting the dataset into testing and training, use Cross-Validation to evaluate the regression performance. What is the RMSE when you use cross validation?

In [81]:
# importing the method:
from sklearn.model_selection import cross_val_score

In [82]:
# Applying 10-fold cross validation with "linear regression":

# In the following line, "my_linreg" is instantiated as an "object" of LinearRegression "class". 
my_linreg = LinearRegression()

mse_list = cross_val_score(my_linreg, X, y, cv=10, scoring='neg_mean_squared_error')

print(mse_list)

# Notice that "cross_val_score" by default provides "negative" values for "mse" to clarify that mse is error.
# in order to calculate root mean square error (rmse), we have to make them positive!
mse_list_positive = -mse_list

# using numpy sqrt function to calculate rmse:
rmse_list = np.sqrt(mse_list_positive)
print(rmse_list)

# calculate the average RMSE as final result of cross validation:
print(rmse_list.mean())

[-23646.90415342 -32003.04401232 -35462.64435619 -37327.60719635
 -14341.32205938 -33628.37104224 -31631.99317834 -12491.00334951
 -20749.61212176 -23204.9474346 ]
[153.7754992  178.89394627 188.31527914 193.20353826 119.75525901
 183.38039983 177.85385343 111.76315739 144.04725656 152.331702  ]
160.33198910744073


#### The RSME with cross validation is higher at 160.33